### ISSUE TITLE
*_path helpers ignore `default_url_options`

### ISSUE THREAD

**Issue by arrtchiu** [rails/rails#27785]
### Steps to reproduce

```ruby
# config/routes.rb
scope '(:paramame)', paramname: /my_regex/, defaults: { paramname: nil } do
  get '/foo', as: :foo
end

# app/controllers/pages_controller.rb
def default_url_options
  if session[:foo] == 'bar'
    { paramname: 'paramvalue' }
  else
    {}
  end
end

# view (haml)
= foo_path
```

### Expected behavior
View outputs `/paramvalue/foo` when `session[:foo] == 'bar'`

### Actual behavior
View outputs '/foo'

### System configuration
**Rails version**: 4.2.7.1  
**Ruby version**: 2.2.2

### Further notes
Modifying `config/routes.rb` to remove `defaults` works:

```ruby
# config/routes.rb
scope '(:paramame)', paramname: /my_regex/ do
  get '/foo', as: :foo
end
```

**Comment by leandro** [comment]
Have you tried this kind of solution?
http://stackoverflow.com/a/5725343/1170086

**Comment by arrtchiu** [comment]
Appreciate you taking the time to suggest a solution, thanks!

Unfortunately this is "global" meaning it's for all requests - certainly this approach works for setting e.g. a host for URLs in mailers, but in this case we really need to determine the values on a per-request basis.

**Comment by arrtchiu** [comment]
I've edited original posting to more accurately reflect the scenario where this reproduces. It seems to be due to the fact that the routes I'm generating paths for are wrapped in a scope that defines defaults. Removing the defaults hash makes it work.

I imagine having both a defaults-hash in `config/routes.rb` and expecting it to honour `default_url_options` would result in a massively complex implementation - instead of changing the behaviour, does it make more sense to consider this a documentation bug?

**Comment by pixeltrix** [comment]
@arrtchiu does your problem go away if you do `foo_path({})` ?

**Comment by coreyward** [comment]
Two questions:

1. Is this typo present in your code, or just the example? `scope '(:paramame)'` (missing 'n')
1. Is `default_url_options` declared as a helper method in the controller? What does `= default_url_options` output in your view?

**Comment by ieldanr** [comment]
Had the same problem when working with locale. `defaults` seems to override `default_url_options` which is totally unexpected.

**Comment by NicosKaralis** [comment]
Same problem as @ieldanr 

My locale set on default_url_options is being ignored

**Comment by arrtchiu** [comment]
@NicosKaralis Check your `config/routes.rb` - there's probably some `defaults` being specified, which causes `default_url_options` not to be called.

**Comment by NicosKaralis** [comment]
@arrtchiu `default_url_options` is being called and it is returning the correct hash values, there is a `defaults` defined in `config/routes.rb` but the shouldn't `default_url_routes` override the default from routes?

```
def default_url_options
  { locale: (I18n.locale == I18n.default_locale ? nil : I18n.locale.to_sym) }
end
```
```
locales = I18n.available_locales - [:en]
scope '(:locale)', locale: /#{locales.join('|')}/, defaults: { locale: I18n.default_locale } do
  get   '/login', to: 'sessions#new', as: :login
  root to: 'home#index'
  match '*a', to: '#missing_action', via: :all, as: nil
end
```

**Comment by arrtchiu** [comment]
> shouldn't default_url_routes override the default from routes?

It seems like it should, but it doesn't. Thinking about how complicated it'd get to support such a thing (should it merge? which one takes precedence?) I don't imagine it'll ever become something supported.

You'll need to remove the defaults from your `scope` block, and allow `en` to be in the regex. That way you can use the `/en/...` routes to allow the user to switch back to English (set it in cookies then redirect back to `/...`).

**Comment by coorasse** [comment]
I double checked and this issue is still present in rails 5.2. I created a repo to reproduce the issue: https://github.com/coorasse/path_helpers_and_defaults and also a small gif to show the error: http://g.recordit.co/d0gOmxhuJk.gif

Basically the `defaults` option in routes overrides the `default_url_options` and this is unexpected.

**Comment by moritzschepp** [comment]
Also, in my case (rails 5.1), its just the generated url helpers that don't work as expected, `url_for` grabs the defaults (only) from `ApplicationController#default_url_options`.

**Comment by everton** [comment]
Still not working in Rails 6.0.2.2. 

Old behavior documented in Ruby guides about how to localize applications was based on the ideia that default_url_options could be override in controllers (allowing for a fine control over generated links):

> https://guides.rubyonrails.org/i18n.html

The content there says: 

> We can include something like this in our ApplicationController then:
>
> # app/controllers/application_controller.rb
> def default_url_options
 > { locale: I18n.locale }
> end

Don't know how to proper circumvent this error.

(An small outburst from an old Rails user: current Rails implementation has 69 modules and ancestor classes to a brand new ApplicationController with 3 different modules just for url_for oO... I'm working with Rails for the last 16 years, but this kind of issue makes me feel that it is becoming impossible to keep using the framework... someone though: "Ohh, ok, this code organization was not what I think is the perfect one" and them move everything into a module, which requires other modules just because yes, and after years of everyone doing this again and again we have 69 modules just to get into a before_action; The same framework which is raising warnings from the language 4 months after a stable version was released... The priorities is clearly strange now days and this is so sad. In the past I advocated a lot about qualities of Rails usage, but it is hard to deal with the current state of the framework development. I know I'm not a contributor and I'm very thankful to all these years I have made my money and joy using the framework, but it is becoming impossible to still on this path :'( )
